_posts/2018-08-10-dota2-macos-performance.md

```diff
@@ -0,0 +1,89 @@
+```bash
+make dota-bench-gl # for GL
+make dota-bench-orig # for MoltenVK bundled with Dota2
+make dota-bench-gfx GFX_METAL_RECORDING=deferred # for gfx-portability with "Deferred" recording
+```
+
+#### Platforms
+
+![macbook fleet](/img/macbook-fleet.jpg)
+
+| Platform         | A | B | C |
+| ---------------- | - | - | - |
+| Model            | MacBook Pro 2016 | MacBook Pro 2016 | MacBook Pro 2013 |
+| CPU              | 3.3 GHz Intel Core i7 | 2.9 GHz Intel Core i7 | 2.6 GHz Intel Core i7 |
+| number of cores  | 2 | 4 | 4
+| GPU              | Intel Iris Graphics 550 | AMD Radeon Pro 460 | NVIDIA GeForce GT 750M |
+| OS               | macOS 10.14 beta | macOS 10.14 beta | macOS 10.13.6 |
+| resolution       | 1440 x 900 | 1680 x 1050 | 1440 x 900 |
+
+### Conclusions
+
+MoltenVK does a good job translating Vulkan to Metal. Interestingly, though, it can be slower than OpenGL on low settings. This doesn't match Phoronix nor Valve/MoltenVK's numbers. We suspect the difference to be explained by Phoronix using a 10x shorter run with pre-loaded pipeline caches. In this case GL would struggle creating all the new pipelines and will not have time to reach the full speed - not exactly an Apples to Apples comparison :)
+
+Strangely, MoltenVK exhibits some sort of "warm-up" behavior, where the second run can be up to 5% faster than the first one (we do include this in the comparison). This is unlike _gfx-portability_, which shows consistent performance. We are not sure what might be causing the warm-up effect, given that we cleared `dota/shadercache/vulkan/shaders.cache` before each run.
+
+Either way, our benchmarks show that OpenGL is still fairly good on MacOS, and it's a high bar to beat for any Vulkan portability implementation, at least in the context of Dota2 performance. The performance shift seen on high settings could be explained by the raised scene complexity (which is where low-level APIs generally help). But, another thing we think could help explain this discrepancy is the significant usage of compute shaders in Dota2, which are not available on MacOS's older OpenGL version. In other words, it's not always the raw performance to gain on MacOS but rather support for more features that allow certain workloads to use the GPU more efficiently.
+
+We believe that immediate command recording has great potential that hasn't yet been realized with Dota2 as it is architectured today with it's one big submission per frame, which increases latency on the already latency-limited program. Hopefully, we'll see more applications taking advantage of this in the future.
+
+#### Rust
+
+Rust has proven itself viable in complex high-performance systems. We were able to build solid abstractions and hide the complexity behind tiny interfaces, while still being able to reason about and optimize low-level performance. Iterating on large architectural changes was a breeze - we just change the core piece and then fix all the compile errors.
```

Comment:
> *the* low-level performance.

(I have a feeling I might've been the one who sugjested [sic] dropping the "the")

Reply:
> we *{could, were able to}* just change 

Also, maybe consider an alternative word to "core"?